### DISCUSSION TITLE
How to cancel client stream?

### DISCUSSION THREAD

**Discussion by sman908** [hyperium/tonic#1268] [Q&A]
Let's say I want to upload a file via grpc stream and IO error happened. How do I cancel the stream?

**Comment by milesj** [comment]
Did you ever figure this out?

**Comment by LucioFranco** [comment]
You would just drop the stream and/or end it.

**Reply by matze** [reply]
Is it possible to cancel the stream by dropping the channel as well? It's awkward that a call (by default) spawns a tokio task that can leak forever if the stream itself is infinite.

**Reply by LucioFranco** [reply]
Good question, I have not tested that and it would be better ask that in the hyper repo since that logic would live there.

**Reply by matze** [reply]
I tested and it is the case (let me know if you want to see reproduction). The easiest mitigation would be to store the `JoinHandle`s in your "executor" (which by the way seems more complicated than necessary) and `abort()` them in the drop handler. But I don't think you want to change the default behaviour?

**Reply by LucioFranco** [reply]
Yeah, that ties us a lot to runtimes, I would say this is an issue with hyper and should be solved there. Since it integrates much closer with the runtime its def the right spot.
